### PR TITLE
Add server-unix-socket option to enable trafic to postgrest server via unix socket

### PR DESCRIPTION
**Pull request by Dansvidania**
Targeting #1186 

Hello, I have corrected the issues from my previous pull request and had to make some changes to make it merge-able after the styling changes

The only thing that is missing is a test in io-test.sh, which I am trying to work on. If anyone is able to give me some pointers on how the bash tests are organized that would be very helpful.

### CODE REVIEW

**Comment by Qu4tro** [comment]
Hey. Sorry about those changes. I'm sure they conflicted with all PRs that were open ahah. 

You can see the function `ensureAppSettings()` on https://github.com/PostgREST/postgrest/blob/master/test/io-tests.sh#L165 . You can basically take that as a template, change the config file for a new one with a `server-unix-socket` key, modify the curl command and check if the results are as expected. I think a simple request should do fine, as the purpose of the test is to determine if the server is reachable and responding in parity to normal usage. Finally, call this new function before `cleanUp` and you should be gtg.

**Comment by Dansvidania** [comment]
No worries actually I also use stylish-haskell and I was worrying I would have to revert all those lines before the PR

I have tried to do just that and then run the circleCi jobs to test, but I am encountering some problems with my bash code (file not found, --unix-socket option not avaliable)

On the other hand, I am having issues running the tests by just running the bash file. I keep getting psql auth bail outs. Am I missing something silly?

**Comment by Dansvidania** [comment]
It seems like the curl version in circleci images is below 7.40 and the --unix-socket option is not available.

what is the preferred way to fix this? would it be ok to download a newer version of curl in the command section on .circleci/config.yaml ?

**Comment by Qu4tro** [comment]
Yep, I don't see any problem with testing with a more recent version of curl.

File: .circleci/config.yml [hunk 1]
```diff
@@ -66,6 +66,11 @@ jobs:
       - restore_cache:
           keys:
           - v1-stack-dependencies-{{ checksum "postgrest.cabal" }}-{{ checksum "stack.yaml" }}
+      - run:
+          name: install ncat
```
Comment (steve-chavez) [inline .circleci/config.yml]: Is `nc` available in CircleCI? Maybe we could omit this dependency, try:

```bash
echo -e "GET / HTTP/1.0\n" | nc -U "/tmp/postgrest.sock"
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -7,6 +7,7 @@ This project adheres to [Semantic Versioning](http://semver.org/).
 
 ### Added
 
+- #1186, Add support for user defined unix socket via server-unix-socket config option
```
Comment (steve-chavez) [inline CHANGELOG.md]: Add surrounding backticks for  `server-unix-socket` so it's more noticeable.

File: test/io-tests.sh [hunk 3]
```diff
@@ -181,6 +181,31 @@ ensureAppSettings(){
   pgrStop
 }
 
+getSocketStatus() {
+  curl -sL -w "%{http_code}\\n" -o /dev/null localhost:54321
```
Comment (steve-chavez) [inline test/io-tests.sh]: If `nc` is available. Then you could do:
```bash
  echo -e "GET / HTTP/1.0\n" | nc -U "/tmp/postgrest.sock" | head -n 1 | awk '{ print $2 }'
```

File: test/io-tests.sh [hunk 4]
```diff
@@ -181,6 +181,31 @@ ensureAppSettings(){
...
+}
+
+socketConnection(){
+  # map port 54321 traffic to unix socket as workaround for old curl version
+  # not supporting --unix-socket flag
+  ncat -vlk 54321 -c 'ncat -U /tmp/postgrest.sock' &
```
Comment (steve-chavez) [inline test/io-tests.sh]: This wouldn't be needed if the above works(`nc`).

File: test/io-tests.sh [hunk 5]
```diff
@@ -181,6 +181,31 @@ ensureAppSettings(){
...
+  pgrStart "./configs/unix-socket.config"
+  while pgrStarted && test "$( getSocketStatus )" -ne 200
+  do
+    # wait for the server to start
+    sleep 0.1 \
+    || sleep 1 # fallback: subsecond sleep is not standard and may fail
+  done
+  sleep 11
```
Comment (steve-chavez) [inline test/io-tests.sh]: `sleep 11` shouldn't be needed if the `while` above is working, try removing it.

File: test/io-tests.sh [hunk 6]
```diff
@@ -181,6 +181,31 @@ ensureAppSettings(){
...
+getSocketStatus() {
+  curl -sL -w "%{http_code}\\n" -o /dev/null localhost:54321
+}
+
+socketConnection(){
+  # map port 54321 traffic to unix socket as workaround for old curl version
+  # not supporting --unix-socket flag
+  ncat -vlk 54321 -c 'ncat -U /tmp/postgrest.sock' &
+  pgrStart "./configs/unix-socket.config"
+  while pgrStarted && test "$( getSocketStatus )" -ne 200
```
Comment (steve-chavez) [inline test/io-tests.sh]: If the change to `nc` is possible, maybe `test -ne` could fail, you could try `test "$( getSocketStatus )" != "200"`

File: .circleci/config.yml (hunk 1 above)
Comment (Dansvidania) [reply .circleci/config.yml]: @steve-chavez I have tried this; for some reason this line works from the command line but I keep getting 400 bad requests when I execute it in a script. 

I have found a reference to [this askubuntu question](https://askubuntu.com/questions/493862/netcat-not-working-at-least-not-that-i-can-tell) but it does not seem to apply for io-tests.sh

any pointers?

Comment (steve-chavez) [reply .circleci/config.yml]: @Dansvidania  Oh. I think it's related to the shell. 

Are you executing the script like `./io-tests.sh`? If so, does it work if you try with `bash io-tests.sh`?

Comment (Dansvidania) [reply .circleci/config.yml]: @steve-chavez unfortunately not.

Comment (steve-chavez) [reply .circleci/config.yml]: @Dansvidania I see. I think for now we can leave the tests with nmap. Check my other comment about `sleep`.

Comment (steve-chavez) [reply .circleci/config.yml]: @Dansvidania Also, please  `rebase` against master. Just merged another PR that changed some of the files. Sorry about that.

**Comment by steve-chavez** [comment]
@Dansvidania Looks good to me!

@Qu4tro What do you think?

**Comment by steve-chavez** [comment]
@Dansvidania Thanks a lot for the work you put on this PR.

I'll merge it now. Soon we'll have a new release :rocket:.

**Comment by Dansvidania** [comment]
@steve-chavez thank you for the feedback. Its a pleasure to be a small part of this great project :)

**Comment by Qu4tro** [comment]
Sorry, been away for a while. This looks great!

Read this the other day that remind me of it: 
https://jvns.ca/blog/2015/11/21/why-you-should-understand-a-little-about-tcp

Haven't actually checked if it's an issue, just something I thought of bringing up :smile:

**Pull request merged**